pkg/virtctl/vnc/vnc.go

```diff
@@ -76,70 +79,103 @@ func (o *VNC) Run(cmd *cobra.Command, args []string) error {
 
+	lnAddr, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
+	if err != nil {
+		return fmt.Errorf("Can't resolve the address: %s", err.Error())
+	}
+
 	// The local tcp server is used to proxy the podExec websock connection to remote-viewer
-	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	ln, err := net.ListenTCP("tcp", lnAddr)
 	if err != nil {
 		return fmt.Errorf("Can't listen on unix socket: %s", err.Error())
 	}
 
 	port := ln.Addr().(*net.TCPAddr).Port
+	var cmnd *exec.Cmd
 
 	// setup connection with VM
 	go func() {
 		err := virtCli.VM(namespace).VNC(vm, pipeInReader, pipeOutWriter)
 		k8ResChan <- err
 	}()
 
+	// wait for remote-viewer to connect to our local proxy server
+	go func() {
+		start := time.Now()
+		glog.Infof("connection timeout: %v", LISTEN_TIMEOUT)
+		// exit early if spawning remote-viewer fails
+		ln.SetDeadline(time.Now().Add(LISTEN_TIMEOUT))
+
+		fd, err := ln.Accept()
+		if err != nil {
+			glog.V(2).Infof("Failed to accept unix sock connection. %s", err.Error())
+			listenResChan <- err
+		}
+		defer fd.Close()
+
+		glog.V(2).Infof("remote-viewer connected in %v", time.Now().Sub(start))
+
+		// write to FD <- pipeOutReader
+		go func() {
+			_, err := io.Copy(fd, pipeOutReader)
+			readStop <- err
+		}()
+
+		// read from FD -> pipeInWriter
+		go func() {
+			_, err := io.Copy(pipeInWriter, fd)
+			writeStop <- err
+		}()
+
+		<-doneChan
+		listenResChan <- err
+	}()
+
 	// execute remote viewer
 	go func() {
-		cmnd := exec.Command("remote-viewer", fmt.Sprintf("vnc://127.0.0.1:%d", port))
-		err := cmnd.Run()
+		args := []string{fmt.Sprintf("vnc://127.0.0.1:%d", port)}
+		if glog.V(4) {
+			args = append(args, "--debug")
+			glog.Infof("remote-viewer commandline: %v", args)
+		}
+
+		cmnd = exec.Command("remote-viewer", args...)
+
+		output, err := cmnd.CombinedOutput()
 		if err != nil {
-			glog.Errorf("remote-viewer execution encountered an error: %v", err)
+			glog.Errorf("remote-viewer execution failed: %v, output: %v", err, string(output))
+		} else {
+			glog.V(2).Infof("remote-viewer output: %v", string(output))
 		}
 		viewResChan <- err
+		close(doneChan)
 	}()
 
-	// wait for remote-viewer to connect to our local proxy server
-	fd, err := ln.Accept()
-	if err != nil {
-		return fmt.Errorf("Failed to accept unix sock connection. %s", err.Error())
-	}
-	defer fd.Close()
-
-	glog.V(2).Infof("remote-viewer connected")
 	go func() {
 		interrupt := make(chan os.Signal, 1)
 		signal.Notify(interrupt, os.Interrupt)
 		<-interrupt
 		close(stopChan)
 	}()
 
-	// write to FD <- pipeOutReader
-	go func() {
-		_, err := io.Copy(fd, pipeOutReader)
-		readStop <- err
-	}()
-
-	// read from FD -> pipeInWriter
-	go func() {
-		_, err := io.Copy(pipeInWriter, fd)
-		writeStop <- err
-	}()
-
 	select {
 	case <-stopChan:
 	case err = <-readStop:
 	case err = <-writeStop:
 	case err = <-k8ResChan:
 	case err = <-viewResChan:
+	case err = <-listenResChan:
 	}
 
+	cmnd.Process.Kill() // avoid leak remote-viewer
```

Comment:
How is that possible? Wouldn't remove-viewer terminate as soon as our proxy server terminates?

Reply:
Not sure how, but during my testing it happened a few time that `virtctl vnc` leaked remote-viewer processes. That *will* eventually terminate, but not before hitting timeout trying to connect to proxy. This could be confusing.

In my case, ti was caused by lack of syncronization between goroutines - this is why I introduced doneChan, which solves the issue

Reply:
ok.

However, changing the cmnd value from nil to an instance in a go-routine is not save. In theory we could get a nil exeception. There is no guarantee that the pointer update will be propagated back to the main routine.

Reply:
Further we probably also want to kill it on signals then ...

Reply:
I decided to remove this code for the time being. Problem is - the issue is too vague. I'm not 100% clear why we leak those processes, so I believe better not commit half-baked code. We can tackle this in another PR once we clearly understand the issue.